api/src/main/java/org/pmiops/workbench/billing/FreeTierBillingService.java

```diff
@@ -298,11 +298,16 @@ private void sendAlertsForCostThresholds(
         workbenchConfigProvider.get().billing.freeTierCostAlertThresholds;
     costThresholdsInDescOrder.sort(Comparator.reverseOrder());
 
+    Iterable<DbUser> userByUserIdIn =
+        userDao.findAllById(userCosts.keySet().stream().collect(Collectors.toList()));
+    Map<Long, DbUser> usersCache = new HashMap<>();
+    userByUserIdIn.iterator().forEachRemaining(u -> usersCache.put(u.getUserId(), u));
+
     userCosts.forEach(
         (userId, currentCost) -> {
           final double previousCost = previousUserCosts.getOrDefault(userId, 0.0);
           maybeAlertOnCostThresholds(
-              userDao.findUserByUserId(userId),
```

Comment:
This is very non-optimal and causes a lot of delay in the job. I'm batching the query to make it faster.

Reply:
this is very similar to the ToS email problem I had.  good update.